PolyEngine/Core/Src/BTreeNode.hpp

```diff
@@ -0,0 +1,605 @@
+#pragma once
+
+#include "Defines.hpp"
+#include "UnsafeStorage.hpp"
+
+namespace Poly {
+	namespace Impl {
+
+		//note(vuko): used to move into uninitialized memory, when using operator=() and (by extension) std::move([range]) is not possible
+		template<typename InIt, typename OutIt>
+		auto constructing_move(InIt first, InIt last, OutIt dest_first) {
+			using InType  = typename std::iterator_traits<InIt>::value_type;
+			using OutType = typename std::iterator_traits<OutIt>::value_type;
+
+			for (; first < last; ++first, ++dest_first) {
+				::new(&*(dest_first)) OutType(std::move(*first));
+				first->~InType();
+			}
+
+			return dest_first;
+		};
+
+		template<typename InputIter, typename DestIter>
+		auto constructing_move_backwards(InputIter first, InputIter last, DestIter dest_last) {
+			using InType  = typename std::iterator_traits<InputIter>::value_type;
+			using OutType = typename std::iterator_traits<DestIter>::value_type;
+
+			while (first < last) {
+				::new(&*(--dest_last)) OutType(std::move(*(--last)));
+				last->~InType();
+			}
+
+			return dest_last;
+		};
+
+		template<typename K, typename V, size_t B>
+		struct BTree {
```

Comment:
I just noticed it. ALL structures and classes need to inherit from BaseObject or (BaseObjectLiteralType in very specific cases). This is major and applies to the whole PR.